### ISSUE TITLE
Ansible version 2 strategy

### ISSUE THREAD

**Issue by leseb** [ceph/ceph-ansible#527]
Dear contributors,

@bmanojlovic ,@hunter ,@guits ,@lorin ,@marmot21 ,@mcsage ,@byronmccollum ,@maethor ,@mhubig ,@laboshinl ,@fcharlier ,@bstillwell ,@lyandrew ,@jjoos ,@BjoernT ,@psy-q ,@bsanders ,@pb-it ,@eikef ,@lpabon ,@ti-mo ,@alfredodeza ,@aisrael ,@andymcc ,@Abhishekvrshny ,@gpocentek ,@Logan2211 ,@git-harry ,@darkcrux ,@nexecook ,@rootfs ,@crcceph ,@mattt416 ,@bengland2 ,@bjne ,@HanXHX ,@andrewschoen ,@matthewrees ,@xals ,@jcftang ,@flisky ,@guestisp ,@msambol

Since Ansible v2 has just been released and wanted your inputs when it comes to the strategy to bring the support for v2 and add more features such as [http://docs.ansible.com/ansible/playbooks_blocks.html](url).

So I have several questions:

* Who moved to v2 already?
* What should be the strategy for v2 in this repo?
I first thought of using branches but they are kinda hard to maintain and require a lot of efforts.

The main issue is that v2 is not 100% backward compatible (even if it claims it is, #523 is a good example for this).

Just to let you know, I'm putting on hold all the PRs that might break `1.9.4`. I'm not sure for how long this version will be supported by Ansible.

Feedback are more than welcome :D.
Thanks!

**Comment by BjoernT** [comment]
While I like to advance on v2 I would suggest so look into the projects using this project. RPC-Openstack is using the ceph roles and that project is still v1 and so does openstack-ansible. Personally I would wait with v2 until all ceph ansible consuming projects moved up to v2

**Comment by leseb** [comment]
Hey @BjoernT, I was actually expected an answer either from you or @mattt416 to get some perspective regarding the integration with both RPC-openstack and openstack-ansible :).

I think that it's wider than 'simply' project consuming ceph-ansible. I'm also thinking of the operators managing their platforms with Ansible. They have other playbooks than ceph-ansible, so even if ceph-ansible becomes v2 ready the rest might not be... Well all of this boils down to the same concern anyway.

The big question is: how and when will we know that we can move to v2? :) For that I'm counting on you for the openstack part at least :).

Thanks!

**Comment by BjoernT** [comment]
I would estimate that we can move to v2 around early June/July this year. I think if we can iron out the bugs to make it v2 compatible and not using v2 features while still maintaining ansible v1 backwards compatibility seems a good compromise to me.

**Comment by Logan2211** [comment]
+1 all of what @BjoernT said. Another downstream openstack-ansible project here with ceph-ansible integration in a similar setup to rpc-openstack. OSA is working on ansible 2 compatibility this cycle and planning to move to it next cycle. Working on the same timeline here and having ceph-ansible compatible with but not requiring v2 for a while would be ideal.

**Comment by jcftang** [comment]
I've been migrating to ansible 2.0 since the dev branch has been around, it's not 100% backwards compatible with ansible 1.9.x.

Also in my environments I usually just install ansible into a virtualenv for specific projects that need older/newer versions.

**Comment by odyssey4me** [comment]
Hi @leseb - For OpenStack-Ansible we're maintaining Ansible 1.9x compatibility for the Mitaka and below releases. We'll only begin working on taking advantage of Ansible 2.x in the next cycle (April 2016).

It'd be ideal if this repo could implement a stable branch strategy, with the stable branch supporting Ansible 1.9x. This would help deployers using OSA 11.x-13.x which will still be supported through until 2017.

Many of our community members would be happy to assist with handling backports and testing them to stable branches.

**Comment by alfredodeza** [comment]
@leseb I know you mentioned that branches would be hard to maintain, but it might be feasible to do so if at a certain point only bugfixes are backported and new features are built onto the 2.0-supporting branch.

I don't think that should be done just yet as it is too early. The schedule being used by openstack-ansible (looking into v2.0 later this year) seems reasonable although I don't think it would be terribly bad to wait for a full transition even later.

**Comment by mattt416** [comment]
Hi @leseb,

Thanks for creating this issue and for including us on it.  I agree w/ @odyssey4me and @alfredodeza that a branching strategy needs to be considered here.  If you did branch on ceph release (for example), you would have to spend some extra cycles managing backports.  However, you gain the ability to ensure that a given ceph release can still be deployed a year down the line while also being able to remove a lot of the release-specific conditional complexity that exists to a degree today.

Thanks!

--Matt

**Comment by git-harry** [comment]
@leseb I agree with @mattt416, you should look to use branches based on the Ceph release cycle. Switching from 1.9 to 2 would then be part of a new release e.g. the next Ceph LTS release.

**Comment by leseb** [comment]
Thank you all for your valuable messages :).

The only thing I'm currently doing with v2 is to implement everything that is backward compatible, so this is mainly syntax things.
Taking advantage of v2 features can definitely wait a couple of months, as mentioned the several openstack ansible projects's timeline sounds reasonable to me.

@alfredodeza I said that maintaining branches would be difficult but I should have said: "difficult for me to handle this on my own :)".
@mattt416 Yes I know and I'm blaming myself a little bit for this. With the infernalis release I spent way to much time handling conditions. Having branches would have been way easier. :)

This week I was planning on tagging the first stable official release of the repo, this idea is to do this under the `1.9.4` branch. This branch will continue to be maintained with additions from Master.
Thus master will likely start implementing new 2.0.0.2 features.

So now, what would you prefer? Having a new branch for Infernalis on `1.9.4` and keep on creating branches for Ceph releases or just a branch for `1.9.4`. I'm just afraid that the release pace of Ceph will not work/sync with what the openstack projects are doing.

**Comment by HanXHX** [comment]
I'm full on ansible v2 except ceph roles.

According with my issue #504, i suggest you to tag. But in long term, you can't maintain 2 versions. 

For example, you can tag and have 2 main branches:
- v1.x --> ansible v1 (current master branch) and freeze development (except bugs)
- v2.x --> ansible v2: new development

You should see how [ansible/ansible](https://github.com/ansible/ansible) works.

**Comment by maethor** [comment]
Thank you very much for including all of us in this issue.

Same as @HanXHX and @jcftang, I use ansible 2 since the release. I'm keeping a virtualenv with ansible 1.9.4 only for ceph. it's really not a big deal, because I'm not deploying a new ceph node every day.

**Comment by aisrael** [comment]
Just to chime in. Where @hunter and I work, we also have a 'downstream' Ansible project that includes `ceph-ansible` as a git submodule. Everything is still predicated on 1.9.4 (we'll probably need to do the same soon: make syntax changes to support 2.0, then move to 2.0 while maintaining backward compatibility, and so on).

How about branching 'off' for Ansible 1.9.x and continuing Ansible 2.0 in `master`? The `ansible-1.9` branch can be barely maintained (only for critical or security updates, maybe), and encourage everyone else to move to Ansible 2.0 and `ceph-ansible:master`?

**Comment by matthewrees** [comment]
I don't have anything to add on what the best strategy might be for future "releases", whether it be branching or tagging for specific Ceph and/or Ansible versions (to allow downstream projects to consume ceph-ansible in a way that best suits them), but as others have mentioned, creating a legacy/1.x branch of the current Infernalis implementation would probably be best.

**Comment by mattt416** [comment]
Hi @leseb,

I certainly think a `1.9.4` branch would be a good start for sure, but do still think it'd be worth considering a more frequent branching strategy in the future.  The `openstack-ansible` project branches on OpenStack release (with master being the development branch), and I feel this works really well because it allows us to do heavy development in master without affecting any of the stable branches.  It also means we can keep our branches relatively lean as they're all self-contained.  Not to mention it feels really good when you can delete a branch and get rid of a bunch of technical debt.  :)

I'm new to Ceph, so I'm not in a great position to say if this strategy would work equally well for ceph-ansible.  For my own clarification, there were only two stable ceph releases in 2015, right?  Looking at http://docs.ceph.com/docs/master/releases/, there are currently 4 stable releases being supported (with Firefly due to be retired soon).  That (with master for development) is quite a lot of branches to maintain, so I suppose that would be something to take into consideration.

--Matt

**Comment by leseb** [comment]
Thanks @mattbenjamin and the others for your inputs again.

The thing is throughout the last Ceph releases I haven't seen much changes.
It is only when Infernalis was released that things changed with a new `ceph` user. Other than that there is nothing. So potentially we could start with the following:

* `1.9.4`: supports giant, hammer and infernalis with ansible 1.9.4 (I don't want to split giant/hammer/infernalis as this will destroy most of my efforts on maintaining the 3 of them in a single branch).
* `2.0.0.2`: supports giant, hammer and infernalis with ansible v2
* `master`: devel branch, new features, only support ansible v2

Then in a couple of months I'll had `2.0.0.2-jewel` along with `1.9.4-jewel` and at the same time `1.9.4` and `2.0.0.2` will disappear. This will allow us to keep on maintaining a branch per Ceph release per Ansible version. `1.9.4` branches won't be getting too much changes but mainly possibles changes for the new Ceph releases where `2.0.0.2` branches will get commits from `master` for new Ansible v2 features and also Ceph releases specific changes.

In terms of naming, I think we should use the ceph release versioning system as at some point the `1.9.4` thing will disappear as we won't be maintaining it forever.
It is kinda messy to have both v1 and v2 for ansible to maintain, at least in terms of naming this is terrible :(
I can't come up with anything really nice, any idea? Current Ceph is `9.2.0` and we need 2 branches for it, so `9.2.0-ansibleV1` and `9.2.0-ansibleV2`? And then how to name releases?

Am I missing something? What do you think?

I'll likely need some help to organise a release process as well. So yes @odyssey4me, I'll be more than happy to get some assistance :).

**Comment by byronmccollum** [comment]
Does Ansible have some internal var / fact that is the Ansible version? If you do come across a incompatible difference between versions, you might be able to handle it via a templated include...

Just a random thought, with no research done on feasibility...

**Comment by byronmccollum** [comment]
This looks promising...

```
ansible infra1 -m debug -a "var=ansible_version"
infra1 | success >> {
    "var": {
        "ansible_version": {
            "full": "1.9.3",
            "major": 1,
            "minor": 9,
            "revision": 3,
            "string": "1.9.3"
        }
    }
}
```

**Comment by leseb** [comment]
@byronmccollum, somehow I had issues in the past using this variable on `1.9.4`.
Indeed this variable could handle both version but not sure if it's the path we want to take...

Anyone? Thoughts?

**Comment by byronmccollum** [comment]
Yeah, it's probably not a good idea long term...

**Comment by odyssey4me** [comment]
@leseb If you're happy to maintain conditionals for the differences between Ceph versions, then perhaps it makes sense to base the branch names on the Ansible versions.

I think that adding more conditionals to have different code paths based on the Ansible version is the road to a lot of pain.

Perhaps the branch names should be something like 'ansible-1.9' and 'ansible-2.0'.

I'm suggesting that the branch include 'ansible' to make it obvious what the version number refers to.
I'm suggesting the use of the major.minor (and not patch) version in the name, as functional breaks will typically happen on the change in minor releases.

I think it's fair to assume that master is the development branch and will converge towards taking advantage of Ansible 2.x as soon as you're ready to do so. I'm not sure that it's worth having an ansible-2.0 branch at this point though - unless you want to manage extra backports, etc.

**Comment by ErwanAliasr1** [comment]
The main question for me is does ceph-ansible is compatible with Ansible 1.9 and 2.0 ? 

In the current situation, the answer is Yes. So nothing in the current tree prevents from using one or the other version of ansible. So @jcftang and @maethor, so you don't need a virtual env for ceph-ansible :).

Creating a 1.9 branch right now means almost managing two base of code and having a clear workflow for it.  I do see several scenario for it.

Scenario 1 aka 1.9 should live long
If a feature is 2.0 specific then get into master and not in 1.9 but what if a feature could work on both ? Shall we include it in both ? The two branches will diverge very fast and will require a lot of work.

Scenario 2 aka 1.9 should die !
The 1.9 branch will only receive important backports but no new feature. That would require few effort to maintain but would expect users to move to 2.0 if they want new features or ceph versions support.

To be honest I really wonder if not creating a branch is such a limitation. As @byronmccollum said, it is possible to detect the version of Ansible and so we can use some "if version" statement in the source code to execute only 2.0 code. What is the price of having a few "if version" inside the source code ?

Until we do have a lot of specific 2.0 code, having everything into the same branch (master) seems reducing the effort to support both releases. I'd suggest considering creating a 1.9 branch like in scenario 2 only once "ansible 1.9 is almost never used" or "we have too much "if version 2" in the code base. This scenario only cost to protect 2.0 code by a "if version 2" statement until we decide there is too many.

My 2 cents,

**Comment by leseb** [comment]
I pretty much like @ErwanAliasr1 idea. I don't much v2 features getting in anytime soon (unless the some syntax, block statement and the package module). So having a master maintaining all of this sounds realistic to me.
Basically, as soon as the if v2 statement requires an include of a large portion of code we will move master's content into a 1.9 branch.

What do you guys think? @mattt416? @odyssey4me? Others?
Thanks!

**Comment by ti-mo** [comment]
@leseb :+1: for that idea, (surprisingly) haven't seen old code break on 2.0 personally, apart from execution being a little slower. Nearly all 2.0 features are clearly marked as such in the Ansible docs, so branching when you want to leave 1.9 behind is fine!

**Comment by odyssey4me** [comment]
@leseb It seems perfectly sensible to me to branch at the latest possible moment, when the conditionals between the two Ansible versions get to a point that it's becoming hard to manage. It would also be best to take the strategy of maintaining 1.9 and 2.x compatibility for as long as possible so that consumers can use either versions. :+1:

**Comment by ErwanAliasr1** [comment]
@leseb Sounds like we have a plan

**Comment by leseb** [comment]
Yes we do! Thank you very much to everyone :D
